client/service/src/config.rs

```diff
@@ -216,11 +206,20 @@ impl<C, G, E> Configuration<C, G, E> {
 	pub fn in_chain_config_dir(&self, sub: &str) -> Option<PathBuf> {
 		self.config_dir.clone().map(|mut path| {
 			path.push("chains");
-			path.push(self.chain_spec.id());
+			path.push(self.expect_chain_spec().id());
 			path.push(sub);
 			path
 		})
 	}
+
+	/// Return a reference to the `ChainSpec` of this `Configuration`.
+	///
+	/// ### Panics
+	///
+	/// This method panic if the `chain_spec` is `None`
+	pub fn expect_chain_spec(&self) -> &ChainSpec<G, E> {
```

Comment:
The idea is to be able to make a `Configuration` before it being modified by the commandline arguments. For that I need defaults and `chain_spec`, like some other fields, are `Option` even though they are required for running a node. IT's not ideal. I think it would be best to have some kind of intermediate object, one that has the `Option`s that can be converted to one that doesn't have and it would be fail-able at that point.

But I don't want to do a complete refactoring. One step at a time. I'm already waiting for too long with this one and I get conflicts every day.

Reply:
This will allow to solve: https://github.com/paritytech/cumulus/issues/24

Either `chain_spec` is not an `Option` but then we need to parse the command line argument to provide a sensible default (which prevents modification of the configuration before parsing the command line), either `chain_spec` is an `Option` but then it means running the node can fail with a runtime error if the developer didn't fill all the required fields in `Configuration`.

Note also that `chain_spec` will not be alone in this situation since `config_dir` has the same issue (if you don't call `init`, it will fail somehow) https://github.com/paritytech/substrate/blob/cecton-cli-refactoring/client/cli/src/lib.rs#L308 and `task_executor` https://github.com/paritytech/substrate/blob/cecton-cli-refactoring/client/service/src/builder.rs#L1152

If I try to revisit and improve the `ServiceBuilder` I will probably fix this. But I think this PR is already big enough.

Reply:
Please create an issue for this!

Reply:
Created #4776